File: scripts/generate-video.py
"""
Veo 3 Video Generation Script (Python)
Generates a video using Gemini's Veo 3 model with image references
"""

import os
import time
from pathlib import Path
from dotenv import load_dotenv
from google import genai

# Load environment variables from .env file
script_dir = Path(__file__).parent
project_root = script_dir.parent
load_dotenv(project_root / '.env')

# Load API key from environment
API_KEY = os.getenv('GOOGLE_GEMINI_API_KEY')
if not API_KEY:
    raise ValueError("GOOGLE_GEMINI_API_KEY environment variable not set")

# Initialize client
client = genai.Client(api_key=API_KEY)

def generate_video():
    print("[*] Starting Veo 3 video generation...\n")

    # Get script directory
    script_dir = Path(__file__).parent
    project_root = script_dir.parent

    # Image paths
    stage_image_path = project_root / "public" / "images" / "anna" / "Anna 1.png"
    podcast_image_path = project_root / "public" / "images" / "anna" / "Anna 2.jpg"

    # Check if images exist
    if not stage_image_path.exists():
        raise FileNotFoundError(f"Stage image not found: {stage_image_path}")
    if not podcast_image_path.exists():
        raise FileNotFoundError(f"Podcast image not found: {podcast_image_path}")

    print(f"[+] Uploading stage image: {stage_image_path.name}")
    stage_file = client.files.upload(file=str(stage_image_path))
    print(f"[+] Uploaded: {stage_file.name}")

    print(f"[+] Uploading podcast image: {podcast_image_path.name}")
    podcast_file = client.files.upload(file=str(podcast_image_path))
    print(f"[+] Uploaded: {podcast_file.name}")

    # Prompt
    prompt = """A professional female expert in her late 30s/early 40s commanding a modern conference stage, speaking passionately about event marketing and communication, warm professional lighting, confident gestures. Camera smoothly transitions to her as a guest on a modern video podcast with contemporary backdrop, engaging in dynamic conversation with host, professional studio setup. Cinematic, empowering, polished. 8 seconds."""

    print(f"\n[*] Prompt: {prompt}\n")
    print("[*] Generating video with Veo 3...")
    print("[*] This may take 1-2 minutes...\n")

    # Generate video with reference images
    # Try using image parameter with multiple images
    operation = client.models.generate_videos(
        model='veo-3.0-generate-001',
        prompt=prompt,
        image={
            'reference_images': [
                {'file_uri': stage_file.uri},
                {'file_uri': podcast_file.uri}
            ]
        }
    )

    print(f"[+] Video generation started!")
    print(f"[*] Operation: {operation.name}\n")

    # Poll for completion
    attempts = 0
    max_attempts = 60  # 5 minutes max

    while not operation.done and attempts < max_attempts:
        attempts += 1
        time.sleep(5)  # Wait 5 seconds

        # Refresh operation status
        operation = client.operations.get(name=operation.name)

        elapsed = attempts * 5
        print(f"[*] Still generating... ({elapsed}s elapsed)", end='\r')

    print()  # New line after progress

    if not operation.done:
        raise TimeoutError("Video generation timed out after 5 minutes")

    if operation.error:
        raise RuntimeError(f"Video generation error: {operation.error}")

    print("[+] Video generation complete!\n")

    # Get the generated video
    video = operation.response.generated_videos[0]

    print("[*] Downloading video...")

    # Download video bytes
    video_data = client.files.download(name=video.video.uri)

    # Save video
    output_dir = project_root / "public" / "videos"
    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / "hero-background.mp4"

    with open(output_path, 'wb') as f:
        f.write(video_data)

    file_size_mb = len(video_data) / (1024 * 1024)

    print(f"[+] Video saved to: {output_path}")
    print(f"[*] Size: {file_size_mb:.2f} MB")

    return str(output_path)

if __name__ == "__main__":
    print("*** Veo 3 Video Generator (Python) ***\n")
    print(f"[*] API Key: {'Present' if API_KEY else 'Missing'}\n")

    try:
        output = generate_video()
        print(f"\n[+] Done! Video saved to: {output}")
    except Exception as e:
        print(f"\n[!] Error: {e}")

        if "404" in str(e) or "not found" in str(e).lower():
            print("\n[!] Veo 3 model not available.")
            print("[*] Make sure:")
            print("   1. Video generation is enabled in Google AI Studio")
            print("   2. Your API key has access to Veo models")

        exit(1)

File: scripts/generate-video.js
/**
 * Veo 3.1 Video Generation Script
 * Generates a video using Gemini's Veo 3.1 model with image reference
 */

import 'dotenv/config';
import { GoogleAIFileManager, GoogleGenerativeAI } from '@google/generative-ai';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

console.log('🚀 Script started...');
console.log('📁 __dirname:', __dirname);
console.log('🔑 API Key present:', !!process.env.GOOGLE_GEMINI_API_KEY);

// Initialize Gemini AI and File Manager
const genAI = new GoogleGenerativeAI(process.env.GOOGLE_GEMINI_API_KEY);
const fileManager = new GoogleAIFileManager(process.env.GOOGLE_GEMINI_API_KEY);

async function generateVideo() {
  try {
    console.log('🎬 Starting Veo 3.1 video generation...');

    // Define the prompt
    const prompt = `A professional female expert in her late 30s/early 40s commanding a modern conference stage, speaking passionately about event marketing and communication, warm professional lighting, confident gestures. Camera smoothly transitions to her as a guest on a modern video podcast with contemporary backdrop, engaging in dynamic conversation with host, professional studio setup. Cinematic, empowering, polished. 8 seconds.`;

    console.log('📝 Prompt:', prompt);

    // Veo 3.1 model
    const model = genAI.getGenerativeModel({ model: 'veo-3.1-generate-preview' });

    // Read reference images (Anna presenting + Anna on podcast)
    const stageImagePath = path.join(__dirname, '../public/images/anna/Anna 1.png');
    const podcastImagePath = path.join(__dirname, '../public/images/anna/Anna 2.jpg');

    let stageImageBase64 = null;
    let podcastImageBase64 = null;

    try {
      const stageBuffer = await fs.readFile(stageImagePath);
      stageImageBase64 = stageBuffer.toString('base64');
      console.log('✅ Stage reference image loaded (Anna 1.png)');
    } catch (err) {
      console.log('⚠️  Stage image not found');
    }

    try {
      const podcastBuffer = await fs.readFile(podcastImagePath);
      podcastImageBase64 = podcastBuffer.toString('base64');
      console.log('✅ Podcast reference image loaded (Anna 2.jpg)');
    } catch (err) {
      console.log('⚠️  Podcast image not found');
    }

    // Build the request parts
    const parts = [
      { text: prompt }
    ];

    // Add reference images if available
    if (stageImageBase64) {
      parts.push({
        inlineData: {
          mimeType: 'image/png',
          data: stageImageBase64
        }
      });
    }

    if (podcastImageBase64) {
      parts.push({
        inlineData: {
          mimeType: 'image/jpeg',
          data: podcastImageBase64
        }
      });
    }

    // Generate video
    console.log('🎥 Generating video with Veo 3.1...');
    console.log('⏳ This may take 1-2 minutes...');

    const result = await model.generateContent({
      contents: [{ role: 'user', parts }],
      generationConfig: {
        temperature: 0.7,
        // Note: Veo 3.1 video parameters might be set in the model config or prompt
        // The API will use default 8 seconds if not specified elsewhere
      }
    });

    const response = result.response;

    // Check if video was generated
    if (response && response.candidates && response.candidates[0]) {
      const candidate = response.candidates[0];

      // Look for video data in the response
      for (const part of candidate.content.parts) {
        if (part.inlineData && part.inlineData.mimeType.startsWith('video/')) {
          const videoData = part.inlineData.data;
          const videoBuffer = Buffer.from(videoData, 'base64');

          // Save video
          const outputPath = path.join(__dirname, '../public/videos/hero-background.mp4');
          await fs.mkdir(path.dirname(outputPath), { recursive: true });
          await fs.writeFile(outputPath, videoBuffer);

          console.log('✅ Video generated successfully!');
          console.log('📁 Saved to:', outputPath);
          console.log('📊 Size:', (videoBuffer.length / 1024 / 1024).toFixed(2), 'MB');

          return outputPath;
        }
      }
    }

    // If we get here, check if there's a file URI or other response
    console.log('📋 Response:', JSON.stringify(response, null, 2));

    throw new Error('No video data found in response');

  } catch (error) {
    console.error('❌ Error generating video:', error.message);

    if (error.message.includes('404') || error.message.includes('not found')) {
      console.error('⚠️  Model not available. Veo 3.1 might not be enabled for your API key yet.');
      console.error('💡 Try:');
      console.error('   1. Check if Veo 3.1 is enabled in your Google AI Studio');
      console.error('   2. Verify you\'re using the correct model name');
      console.error('   3. Check API quotas and billing');
    }

    throw error;
  }
}

// Run if called directly
const isMainModule = import.meta.url.endsWith(process.argv[1].replace(/\\/g, '/'));

console.log('🔍 Checking if main module...');
console.log('   import.meta.url:', import.meta.url);
console.log('   process.argv[1]:', process.argv[1]);
console.log('   isMainModule:', isMainModule);

if (isMainModule || process.argv[1].includes('generate-video.js')) {
  console.log('✅ Running as main module, starting generation...');
  generateVideo()
    .then(() => {
      console.log('🎉 Video generation complete!');
      process.exit(0);
    })
    .catch((error) => {
      console.error('💥 Fatal error:', error);
      process.exit(1);
    });
}

export { generateVideo };
